### PR TITLE
PCI device passthrough option for vm-manager

### PR DESCRIPTION
**Pull request by QiaosongZhou**
First commit for pci-passthrough feature extension for vm-manager.
Added graphical feature to allow multiple selections in a suboption menu box.
Added secure vfio setup for passthrough devices.

Tracked-On: OAM-95606
Signed-off-by: Qiaosong Zhou <qiaosong.zhou@intel.com>

### CODE REVIEW

File: src/guest/guest.c [hunk 1]
```diff
@@ -121,6 +122,29 @@ int load_form_data(char *name)
 	val = g_key_file_get_string(in, g->name, g->key[EXTRA_CMD], NULL);
 	set_field_data(FORM_INDEX_EXTRA_CMD, val);
 
+	g = &g_group[GROUP_PCI_PT];
+	val = g_key_file_get_string(in, g->name, g->key[PCI_PT], NULL);
+	set_field_data(FORM_INDEX_PCI_PT, val);
+	// val = g_key_file_get_string(in, g->name, g->key[PCI_PT_USB], NULL);
```
Comment (YadongQi) [inline src/guest/guest.c:128]: remove deprecated code

**Review by YadongQi — CHANGES_REQUESTED** [review]

File: src/guest/guest.c [hunk 2]
```diff
@@ -31,6 +31,7 @@ keyfile_group_t g_group[] = {
 	{ "vtpm",     { "bin_path", "data_dir", NULL } },
 	{ "rpmb",     { "bin_path", "data_dir", NULL } },
 	{ "extra",    { "cmd", NULL } },
+	{ "passthrough", { "passthrough_pci", "usb", "udc", "audio", "eth", "wifi", NULL}},
```
Comment (YadongQi) [inline src/guest/guest.c:34]: were the "usb", "udc" ... still usefull?

File: src/guest/tui.c [hunk 3]
```diff
@@ -206,6 +226,16 @@ static int popup_sub_menu(WINDOW *parent, const char **opts, int nb_opts)
 	return ret;
 }
 
+// static pop_up_select_menu(WINDOW *parent, const char **opts, int nb_opts, const int selected[]) {
```
Comment (YadongQi) [inline src/guest/tui.c:229]: remove?

File: src/guest/tui.c [hunk 4]
```diff
@@ -532,6 +646,10 @@ static int form_key_enter(void)
 					field_opts_off(g_form_field_data[FORM_INDEX_VGPU_GVTG_UUID].input.f, O_EDIT);
 				}
 			}
+		} else if (ff->input.sub_opts->selected) {
+			if (form->current == g_form_field_data[FORM_INDEX_PCI_PT].input.f) {
+
```
Comment (YadongQi) [inline src/guest/tui.c:651]: empty?

File: src/guest/tui.c [hunk 5]
```diff
@@ -645,22 +763,27 @@ int get_field_data(form_index_t index, char *out, size_t out_len)
...
+	} else {
+		strncpy(out,field_buffer(g_form_field_data[index].input.f, 0),out_len);
 	}
 
 	g_strstrip(out);
 
 	return 0;
 }
 
+
```
Comment (YadongQi) [inline src/guest/tui.c:786]: why add this new line? remove it

File: src/include/guest/guest.h [hunk 6]
```diff
@@ -87,6 +88,22 @@ enum {
...
+	PCI_PT_OPTS_AUDIO,
+	PCI_PT_OPTS_ETH,
+	PCI_PT_OPTS_WIFI
+};
+
+#define PCI_PT_OPTS_USB_STR "usb"
+#define PCI_PT_OPTS_UDC_STR "udc"
+#define PCI_PT_OPTS_AUDIO_STR "audio"
+#define PCI_PT_OPTS_ETH_STR "eth"
+#define PCI_PT_OPTS_WIFI_STR "wifi"
```
Comment (YadongQi) [inline src/include/guest/guest.h:105]: remove these define if useless

File: src/utils.c [hunk 7]
```diff
@@ -112,3 +120,433 @@ int cleanup_child_proc(void)
...
+		close(fd);
+		return -1;
+	}
+
+	close(fd);
+
+	return 0;
+}
+
+void print_regerror (int errcode, size_t length, regex_t *compiled)
```
Comment (YadongQi) [inline src/utils.c:434]: take care the coding styles for below codes

**Comment by QiaosongZhou** [comment]
Completed requested changes, please verify.

File: src/utils.c [hunk 8]
```diff
@@ -119,3 +119,436 @@ int cleanup_child_proc(void)
...
+	}
+	return count_res;
+}
+
+
+int find_ProgIf(char *pci_device, char * res) {
+	/* returns 0 if ProgIf is found, -1 if not */
+	char buffer[4096];
+	// char pci_device[16] = {0};
+    char search_key[64] = "ProgIf";
```
Comment (YadongQi) [inline src/utils.c:506]: tab as indent

File: src/utils.c [hunk 9]
```diff
@@ -119,3 +119,436 @@ int cleanup_child_proc(void)
...
+		return -1;
+	}
+	return count_res;
+}
+
+
+int find_ProgIf(char *pci_device, char * res) {
+	/* returns 0 if ProgIf is found, -1 if not */
+	char buffer[4096];
+	// char pci_device[16] = {0};
```
Comment (YadongQi) [inline src/utils.c:505]: remove if useless

File: src/utils.c [hunk 10]
```diff
@@ -119,3 +119,436 @@ int cleanup_child_proc(void)
...
+	{
+		if (result == REG_ESPACE)
+			fprintf (stderr, "%s\n", strerror(ENOMEM));
+		else
+			fputs ("Syntax error in the regular expression passed as first argument\n", stderr);
+		return -1;               
+	}
+
+	regmatch_t matches[1];
+    int count_res = 0;
```
Comment (YadongQi) [inline src/utils.c:521]: ditto

File: src/utils.c [hunk 11]
```diff
@@ -119,3 +119,436 @@ int cleanup_child_proc(void)
...
+	/* Open the command for reading. */
+    char *options[] = {"lspci", "-vmms", pci_device, (char *)0};
```
Comment (YadongQi) [inline src/utils.c:523]: ditto

File: src/utils.c [hunk 12]
```diff
@@ -119,3 +119,436 @@ int cleanup_child_proc(void)
...
+    char *env[] = {(char *)0};
```
Comment (YadongQi) [inline src/utils.c:524]: ditto

File: src/utils.c [hunk 13]
```diff
@@ -119,3 +119,436 @@ int cleanup_child_proc(void)
...
+
+void print_regerror(int errcode, size_t length, regex_t *compiled)
+{
+	char buffer[length];
+	(void) regerror (errcode, compiled, buffer, length);
+	fprintf(stderr, "Regex match failed: %s\n", buffer);
+}
+
+int find_pci(const char * name, int n, char *res[]) {
+    /* returns number of pci devices found, -1 if error. */
```
Comment (YadongQi) [inline src/utils.c:444]: tab as indent

File: src/utils.c [hunk 14]
```diff
@@ -119,3 +119,436 @@ int cleanup_child_proc(void)
...
+	char buffer[4096];
+	// char pci_device[16] = {0};
+    char search_key[64];
```
Comment (YadongQi) [inline src/utils.c:447]: ditto

File: src/utils.c [hunk 15]
```diff
@@ -119,3 +119,436 @@ int cleanup_child_proc(void)
...
+	{
+		if (result == REG_ESPACE)
+			fprintf (stderr, "%s\n", strerror(ENOMEM));
+		else
+			fputs ("Syntax error in the regular expression passed as first argument\n", stderr);
+		return -1;               
+	}
+
+	regmatch_t matches[1];
+    int count_res = 0;
```
Comment (YadongQi) [inline src/utils.c:462]: ditto, please check again with below lines

File: src/utils.c [hunk 16]
```diff
@@ -14,7 +14,15 @@
 #include <errno.h>
```
Comment (YadongQi) [inline src/utils.c:14]: add "Tracked-On: OAM-95606" in commit message

File: src/include/utils.h [hunk 17]
```diff
@@ -8,8 +8,29 @@
 #ifndef __UTILS_H__
 #define __UTILS_H__
 
+#include <sys/types.h>
+#include <regex.h>
```
Comment (YadongQi) [inline src/include/utils.h:12]: remove these header include, move them to c file.

File: src/guest/start.c [hunk 18]
```diff
@@ -174,6 +182,173 @@ static int passthrough_gpu(void)
...
+			if (strcmp(pci_device, ".") == 0 || strcmp(pci_device, "..") == 0) {
+				continue;
+			}
+			char buffer[128] = {0}, device[64] = {0}, vendor[64] = {0};
+			
+			/* Get device_id and vendor_id */
+			snprintf(buffer, 128, "%s/%s/device", iommu_grp_dev, pci_device);
+			fprintf(stderr, "%s\n", buffer);
+			FILE *f = fopen(buffer, "r");
+			if (f == NULL && !fgets(device, sizeof(device), f)) {
```
Comment (YadongQi) [inline src/guest/start.c:276]: f == NULL || !fgets(device, sizeof(device), f)

File: src/guest/start.c [hunk 19]
```diff
@@ -390,27 +565,57 @@ int start_guest(char *name)
...
+		for (int i=0; i<res_count; i++) {
+			if (strcmp(ptr, temp[i]) == 0) {
+				cx = setup_passthrough_pci(ptr, p, cx);
+				p += cx; size -= cx;
+				break;
+			}
+		}
+		ptr = strtok(NULL, delim);
+	}
+
```
Comment (YadongQi) [inline src/guest/start.c:596]: move these codes between line-554 and line-555.

**Review by YadongQi — APPROVED** [review]

**Review by YadongQi — APPROVED** [review]

**Review by YadongQi — APPROVED** [review]

**Review by YadongQi — APPROVED** [review]

**Review by YadongQi — APPROVED** [review]

**Pull request merged**